File: examples/timing/mxm.c
#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "mpi.h"
#include <string.h>

#include "blaslapack.h"
#include "common.h"

void mxm_std (Matrix a, Matrix b, Matrix c)
{
  int i, j, l;
  for (i=0; i <  a->rows; i++) {
    for (j=0; j  < b->cols; j++) {
      c->data[i][j] =  0.0;
      for (l=0;  l < a->cols; l++) {
        c->data[j][i] += a->data[l][i]*b->data[j][l];
      }
    }
  }
}

void mxm_unr (Matrix a, Matrix b, Matrix c)
{
  int i, j;
  for (i=0; i < a->rows; i++) {
    for (j=0; j < b->cols; j++) {
      c->data[i][j] = a->data[i][0]*b->data[0][j]
        +a->data[1][i]*b->data[j][1]
        +a->data[2][i]*b->data[j][2]
        +a->data[3][i]*b->data[j][3]
        +a->data[4][i]*b->data[j][4]
        +a->data[5][i]*b->data[j][5]
        +a->data[6][i]*b->data[j][6]
        +a->data[7][i]*b->data[j][7]
        +a->data[8][i]*b->data[j][8]
        +a->data[9][i]*b->data[j][9];
    }
  }
}

void mxm_blas (Matrix a, Matrix b, Matrix c)
{
  char    trans = 'N';
  double alpha = 1.0, beta = 0.0;
  dgemm(&trans, &trans, &a->rows, &b->cols, &a->cols, &alpha,
        a->data[0], &a->cols, b->data[0], &a->rows, &beta, c->data[0], &a->rows);
}

void mxm (Matrix a, Matrix b, Matrix c, int flag)
{
  if (flag == 0)
    mxm_std(a, b, c); 
  else if (flag == 1 && a->cols == 10)
    mxm_unr(a, b, c); 
  else if (flag == 2)
    mxm_blas(a, b, c);
  else {
    printf (" Incorrect setting of variable flag in mxm \n");
    close_app();
    exit(1);
  }
}

int main(int argc, char **argv )
{
  int size, rank;
  long i, j, l, m, n, k, iter, loop;
  Matrix a, b, c;
  double t1, t2, dt, dt1, r;

  init_app(argc, argv, &rank, &size);

  if( argc < 3 ) {
    if (rank == 0)
      printf("need atleast 2 arguments - n & flag\n");
    close_app();
    return 1;
  }

  k = m = n = atoi(argv[1]);
  a = createMatrix(m, k);
  b = createMatrix(k, n);
  c = createMatrix(m, n);

  for (i=0; i  < m; i++) {
    for (l=0; l < k; l++) {
      a->data[l][i] = i+1.0;
    }
  }
  for (l=0; l  < k; l++) {
    for (j=0; j < n; j++) {
      b->data[j][l] = j+1.0;
    }
  }
  loop = 5;

  t1 = WallTime(); 
  for (iter=0; iter < loop; iter++)
    mxm (a,b,c,atoi(argv[2]));
  t2 = WallTime();

  dt = t2 - t1;
  dt1 = dt/(m*2*k*n);
  dt1 = dt1/loop;
  r    = 1.e-6/dt1;
  printf (" matrix-matrix : (m,k,n)= (%ld,%ld,%ld)    dt= %lf (s) dt1= %le r= %lf\n" ,m, k, n, dt, dt1, r);

  freeMatrix(a);
  freeMatrix(b);
  freeMatrix(c);

  close_app();

  return 0;
}

File: examples/timing/mxm.f
      program mxmtest
      include 'mpif.h'
      parameter (m = @SIZE@); 
      parameter (k = @SIZE@); 
      parameter (n = @SIZE@); 
      real*8 a(m,k), b(k,n), c(m,n), t1, t2, dt, dt1, r
      integer*8 loop, m8, k8, n8

      m8 = m;
      k8 = k;
      n8 = n;

      call MPI_INIT (ierr)
      call MPI_COMM_RANK (MPI_COMM_WORLD, myid, ierr)
      call MPI_COMM_SIZE (MPI_COMM_WORLD, nproc, ierr)
      do l=1,k
         do i=1,m
            a(i,l) = dble(i+1)
         enddo
      enddo
      do j=1,n
         do l=1,k
            b(l,j) = dble(j+1)
         enddo
      enddo
      loop = 5
      t1 = MPI_WTIME(ierr)
      do iter=1,loop
          call mxm (a,m,b,k,c,n)
      enddo
      t2 = MPI_WTIME(ierr)
      dt = t2 - t1
      dt1 = dt/(m8*2*k8*n8)
      dt1 = dt1/loop
      r   = 1.E-6/dt1
      write (6,*) 'matrix-matrix :(m,k,n)= ',m,k,n
      write (6,*) 'r=',r
      call MPI_FINALIZE(ierr)
      stop
      end

      subroutine mxm(a,m,b,k,c,n)
c-------------------------------------------------------------
c matrix-matrix product
c      = a*b
c-------------------------------------------------------------
      real*8 a(m,k), b(k,n),c(m,n)
      real*8 alpha, beta
      integer flag
      flag = @FLAG@
      if (flag .eq. 0) then
         call mxm_std(a,m,b,k,c,n)
      else if (flag .eq. 1) then
        call mxm_unr(a,m,b,k,c,n)
      else if (flag .eq. 2) then
          alpha = 1.0
          beta = 0.0
          call dgemm('N','N',m,n,k,alpha,a,m,b,k,beta,c,m)
      else
          write(6,*)'incorrect flag in mxm'
          stop
      endif
      return
      end

      subroutine mxm_std(a,m,b,k,c,n)
c---------------------------------------------------------
c matrix-matrix product
c = a*b
c "standard" case
c---------------------------------------------------------
      real*8 a(m,k), b(k,n), c(m,n)
      do j=1,n
        do i=1,m
            c(i,j) = 0.0
            do l=1,k
                c(i,j) = c(i,j) + a(i,l)*b(l,j)
            enddo
         enddo
      enddo
      return
      end
      
      subroutine mxm_unr(a,m,b,k,c,n)
c---------------------------------------------------------
c matrix-matrix product
c = a*b
c unrolled case (k=10)
c---------------------------------------------------------
      real*8 a(m,k), b(k,n), c(m,n)
      do j=1,n
        do i=1,m
            c(i,j) = a(i,1)*b(1,j)
     +            + a(i,2)*b(2,j)
     +            + a(i,3)*b(3,j)
     +            + a(i,4)*b(4,j)
     +            + a(i,5)*b(5,j)
     +            + a(i,6)*b(6,j)
     +            + a(i,7)*b(7,j)
     +            + a(i,8)*b(8,j)
     +            + a(i,9)*b(9,j)
     +            + a(i,10)*b(10,j)
        enddo
      enddo
      return
      end
